File: demo/sign.php
<!-- $Id: sign.php 1104 2017-12-06 13:58:03Z kgoldman $ -->

<?php
/* (c) Copyright IBM Corporation 2016.						*/
/*										*/
/* All rights reserved.								*/
/* 										*/
/* Redistribution and use in source and binary forms, with or without		*/
/* modification, are permitted provided that the following conditions are	*/
/* met:										*/
/* 										*/
/* Redistributions of source code must retain the above copyright notice,	*/
/* this list of conditions and the following disclaimer.			*/
/* 										*/
/* Redistributions in binary form must reproduce the above copyright		*/
/* notice, this list of conditions and the following disclaimer in the		*/
/* documentation and/or other materials provided with the distribution.		*/
/* 										*/
/* Neither the names of the IBM Corporation nor the names of its		*/
/* contributors may be used to endorse or promote products derived from		*/
/* this software without specific prior written permission.			*/
/* 										*/
/* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS		*/
/* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT		*/
/* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR	*/
/* A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT		*/
/* HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,	*/
/* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT		*/
/* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,	*/
/* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY	*/
/* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT		*/
/* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE	*/
/* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.		*/
/********************************************************************************/
?>

<html>
<head>
<title>TSS 2.0 Demo Sign
<?php
echo gethostname();
?>
</title>
<link rel="stylesheet" type="text/css" href="demo.css">
</head>
			
<div id="header">
<img src="ibm.png" style="float:right;width:200px;height:70px">
<h2>IBM TSS Demo RSA Sign and Verify - 
<?php
echo gethostname();
?>
</h2>
</div>

<?php
require '/var/www/html/tpm2/nav.html';
?>

<div id="section">

<?php
if ( $_SERVER['REQUEST_METHOD'] == 'POST' ) {
    //print_r($_POST);
    echo "<br>";

    $command = $_POST['command'];
    $hp = $_POST['hp'];
    $label= $_POST['label'];
    $sigfile= $_POST['sigfile'];
    $msg = $_POST['msg'];
    $pwdk = $_POST['pwdk'];

    $retval = 0;
    // parameter checks
    if ($retval == 0) {
	if (strlen($hp) == 0) {
	    echo "Parent handle must be specified<br>";
	    $retval = 1;
	}
	if (strlen($label) == 0) {
	    echo "Label must be specified<br>";
	    $retval = 1;
	}
	if (strlen($sigfile) == 0) {
	    echo "Signature name must be specified<br>";
	    $retval = 1;
	}
 	if (strlen($msg) == 0) {
	    echo "Message must be specified<br>";
	    $retval = 1;
	}
    }
    if ($retval == 0) {
	$rc = file_put_contents ('message.tmp', $msg);
	if (!$rc) {
	    echo "could not write message to message.tmp<br>";
	    $retval = 1;
	}
    }
    // load the key
    if ($retval == 0) {
        $commandStr = "/var/www/html/tpm2/load";
        $commandStr .= " -hp " . $hp;
        $commandStr .= " -ipu " . $label . "pub.key";
        $commandStr .= " -ipr " . $label . "priv.key";
        if (strlen($pwdp) != 0) {
            $commandStr .= " -pwdp " . $pwdp;
        }
        //echo 'Command string: ' . $commandStr. "<br>";
        unset($output);
        exec ($commandStr, $output, $retval);
        if ($retval != 0) {
            echo $commandStr . "<br>";
            for ($i = 0 ; $i < count($output) ; $i++) {
                echo $output[$i] . "<br>";
            }
        }
    }
    // get the handle from the response
    if ($retval == 0) {
        //print_r($output);
        $values = explode (" ", $output[0]);
        $hk = $values[1];
        echo "Loaded signing key handle: " . $hk . "<br>";
    }
    // construct the sign or verify command using the signing key
    if ($retval == 0) {
	switch ($command) {
	  case 'Sign':
	    $commandStr = "/var/www/html/tpm2/sign";
	    $commandStr .= " -hk " . $hk;
	    $commandStr .= " -os " . $sigfile . ".sig";
	    $commandStr .= " -if message.tmp";
	    if (strlen($pwdk) != 0) {
	        $commandStr .= " -pwdk " . $pwdk;
	    }
	    break;
	  case 'Verify Signature':
	    $commandStr = "/var/www/html/tpm2/verifysignature";
	    $commandStr .= " -hk " . $hk;
	    $commandStr .= " -is " . $sigfile . ".sig";
	    $commandStr .= " -if message.tmp";
	    break;
	  default:
	    echo ("Invalid command $command");
	    $retval = 1;
	    break;
	}
    }
    // run the sign or verify command
    if ($retval == 0) {
	//echo 'Command string: ' . $commandStr. "<br>";
	unset($output);
	exec ($commandStr, $output, $retval);
	if ($retval == 0) {
	    if ($command == 'Sign') {
		;
	    }
	    else if ($command == 'Verify Signature') {
		;
	    }
	    echo "Success";
	}
	else {
	    echo $commandStr . "<br>";
	    for ($i = 0 ; $i < count($output) ; $i++) {
		echo $output[$i] . "<br>";
	    }
	}
    }
    unlink ('message.tmp');
    // flush  
    if (strlen($hk) != 0) {
        $commandStr = "/var/www/html/tpm2/flushcontext";
        $commandStr .= " -ha " . $hk;
        //echo 'Command string: ' . $commandStr. "<br>";
        unset($output);
        exec ($commandStr, $output, $retval);
        if ($retval != 0) {
            echo $commandStr . "<br>";
            for ($i = 0 ; $i < count($output) ; $i++) {
                echo $output[$i] . "<br>";
            }
        }
    }
}
?>

<form method="post" action="sign.php">

<h3>Parameters</h3>
     
Parent Handle <input type="text" name="hp" value="<?php
if (strlen($hp) != 0) {
echo $hp;
}
 else {
echo "80000000";
 }
?>"><br>
Key Label <input type="text" name="label" value="<?php echo $label; ?>">
<br>
Message <input type="text" name="msg" value="<?php echo $msg; ?>">
<br>
Signature Name <input type="text" name="sigfile" value="<?php echo $sigfile; ?>">
<br>
     
<h3>Sign</h3>

Key Password <input type="password" name="pwdk" value="<?php echo $pwdk; ?>">
<br>
<input type="submit" name="command" value="Sign">

<h3>Verify</h3>
<input type="submit" name="command" value="Verify Signature">

</div>

<?php
require '/var/www/html/tpm2/footer.html';
?>

</body>
</html>

File: utils12/sign.c
/********************************************************************************/
/*										*/
/*			    TPM 1.2 Sign					*/
/*			     Written by Ken Goldman				*/
/*		       IBM Thomas J. Watson Research Center			*/
/*	      $Id: sign.c 1286 2018-07-27 19:20:16Z kgoldman $		*/
/*										*/
/* (c) Copyright IBM Corporation 2018.						*/
/*										*/
/* All rights reserved.								*/
/* 										*/
/* Redistribution and use in source and binary forms, with or without		*/
/* modification, are permitted provided that the following conditions are	*/
/* met:										*/
/* 										*/
/* Redistributions of source code must retain the above copyright notice,	*/
/* this list of conditions and the following disclaimer.			*/
/* 										*/
/* Redistributions in binary form must reproduce the above copyright		*/
/* notice, this list of conditions and the following disclaimer in the		*/
/* documentation and/or other materials provided with the distribution.		*/
/* 										*/
/* Neither the names of the IBM Corporation nor the names of its		*/
/* contributors may be used to endorse or promote products derived from		*/
/* this software without specific prior written permission.			*/
/* 										*/
/* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS		*/
/* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT		*/
/* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR	*/
/* A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT		*/
/* HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,	*/
/* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT		*/
/* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,	*/
/* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY	*/
/* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT		*/
/* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE	*/
/* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.		*/
/********************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <ibmtss/tss.h>
#include <ibmtss/tssutils.h>
#include <ibmtss/tssresponsecode.h>
#include <ibmtss/tsscryptoh.h>
#include <ibmtss/tsscrypto.h>
#include <ibmtss/tpmstructures12.h>
#include <ibmtss/tssmarshal12.h>
#include <ibmtss/Unmarshal12_fp.h>
#include "cryptoutils.h"

static void printUsage(void);

int verbose = FALSE;

int main(int argc, char * argv[])
{
    TPM_RC 			rc = 0;
    int				i;				/* argc iterator */
    TSS_CONTEXT			*tssContext = NULL;
    Sign12_In			in;
    Sign12_Out			out;
    TPM_KEY_HANDLE		keyHandle = 0;
    const char			*keyPassword = NULL; 
    const char			*signatureFilename = NULL;
    const char			*inputFilename = NULL;
    unsigned char 		*input = NULL;
    size_t 			inputlength;
    TPMT_HA 			areaToSign;
    const char 			*keyFilename = NULL;
    TPM_AUTHHANDLE 		sessionHandle0 = TPM_RH_NULL;
    unsigned int		sessionAttributes0 = 0;
	
    setvbuf(stdout, 0, _IONBF, 0);      /* output may be going through pipe to log file */
    TSS_SetProperty(NULL, TPM_TRACE_LEVEL, "1");

    for (i=1 ; (i<argc) && (rc == 0) ; i++) {
	if (strcmp(argv[i],"-hk") == 0) {
	    i++;
	    if (i < argc) {
		sscanf(argv[i],"%x", &keyHandle);
	    }
	    else {
		printf("Missing parameter for -hk\n");
		printUsage();
	    }
	}
	else if (!strcmp("-pwdk",argv[i])) {
	    i++;
	    if (i < argc) {
		keyPassword = argv[i];
	    } else {
		printf("Missing parameter for -pwdk\n");
		printUsage();
	    }
	}
	else if (strcmp(argv[i],"-if") == 0) {
	    i++;
	    if (i < argc) {
		inputFilename = argv[i];
	    }
	    else {
		printf("-if option needs a value\n");
		printUsage();
	    }
	}
	else if (strcmp(argv[i],"-os") == 0) {
	    i++;
	    if (i < argc) {
		signatureFilename = argv[i];
	    }
	    else {
		printf("-os option needs a value\n");
		printUsage();
	    }
	}
	else if (strcmp(argv[i],"-ik") == 0) {
	    i++;
	    if (i < argc) {
		keyFilename = argv[i];
	    }
	    else {
		printf("-ik option needs a value\n");
		printUsage();
	    }
	}
	else if (strcmp(argv[i],"-se0") == 0) {
	    i++;
	    if (i < argc) {
		sscanf(argv[i],"%x", &sessionHandle0);
	    }
	    else {
		printf("Missing parameter for -se0\n");
		printUsage();
	    }
	    i++;
	    if (i < argc) {
		sscanf(argv[i],"%x", &sessionAttributes0);
		if (sessionAttributes0 > 0xff) {
		    printf("Out of range session attributes for -se0\n");
		    printUsage();
		}
	    }
	    else {
		printf("Missing parameter for -se0\n");
		printUsage();
	    }
	}
	else if (!strcmp(argv[i], "-h")) {
	    printUsage();
	}
	else if (!strcmp(argv[i], "-v")) {
	    verbose = TRUE;
	    TSS_SetProperty(NULL, TPM_TRACE_LEVEL, "2");
	}
	else {
	    printf("\n%s is not a valid option\n", argv[i]);
	    printUsage();
	}
    }
    if (keyHandle == 0) {
	printf("Missing handle parameter -hk\n");
	printUsage();
    }
    if (inputFilename == NULL) {
	printf("Missing input filename -if\n");
	printUsage();
    }
    if (rc == 0) {
	rc = TSS_File_ReadBinaryFile(&input,     	/* freed @1*/
				     &inputlength,
				     inputFilename);
    }
    if (rc == 0) {
	in.keyHandle = keyHandle;
	areaToSign.hashAlg = TPM_ALG_SHA1; 
	rc = TSS_Hash_Generate(&areaToSign,
			       inputlength, input,
			       0, NULL);
	memcpy(in.areaToSign, (uint8_t *)&areaToSign.digest, SHA1_DIGEST_SIZE);
	in.areaToSignSize = SHA1_DIGEST_SIZE;
    }
    /* Start a TSS context */
    if (rc == 0) {
	rc = TSS_Create(&tssContext);
    }
    if (rc == 0) {
	rc = TSS_Execute(tssContext,
			 (RESPONSE_PARAMETERS *)&out,
			 (COMMAND_PARAMETERS *)&in,
			 NULL,
			 TPM_ORD_Sign,
			 sessionHandle0, keyPassword, sessionAttributes0,
			 TPM_RH_NULL, NULL, 0);
    }
    {
	TPM_RC rc1 = TSS_Delete(tssContext);
	if (rc == 0) {
	    rc = rc1;
	}
    }
    if ((rc == 0) && (signatureFilename != NULL)) {
	rc = TSS_File_WriteBinaryFile((uint8_t *)out.sig, out.sigSize, signatureFilename) ;
    }
    /* if a key file was specified, verify the signature */
    if (keyFilename != NULL) {
	TPM_KEY12 	signingKey;
	RSA         	*rsaPubKey = NULL;
	TPMT_SIGNATURE 	tSignature;

	/* get the signing key */
	if (rc == 0) {
	    rc = TSS_File_ReadStructure(&signingKey,
					(UnmarshalFunction_t)TSS_TPM_KEY12_Unmarshalu,
					keyFilename);
	}
	/* construct the OpenSSL RSA public key token */
	if (rc == 0) {
	    unsigned char earr[3] = {0x01, 0x00, 0x01};
	    rc = TSS_RSAGeneratePublicToken
		 (&rsaPubKey,			/* freed @3 */
		  signingKey.pubKey.key,	 	/* public modulus */
		  signingKey.pubKey.keyLength,
		  earr,      			/* public exponent */
		  sizeof(earr));
	}
	if (rc == 0) {
	    rc = convertRsaBinToTSignature(&tSignature,
					   TPM_ALG_SHA1,
					   out.sig,
					   out.sigSize);
	}
	/* verify the TPM signature */
	if (rc == 0) {
	    rc = verifyRSASignatureFromRSA((uint8_t *)&areaToSign.digest,
					   SHA1_DIGEST_SIZE,
					   &tSignature,
					   TPM_ALG_SHA1,
					   rsaPubKey);
	}
	if (rsaPubKey != NULL) {
	    RSA_free(rsaPubKey); 	/* @3 */
	}
    }
    if (rc == 0) {
	if (verbose) printf("sign: success\n");
    }
    else {
	const char *msg;
	const char *submsg;
	const char *num;
	printf("sign: failed, rc %08x\n", rc);
	TSS_ResponseCode_toString(&msg, &submsg, &num, rc);
	printf("%s%s%s\n", msg, submsg, num);
	rc = EXIT_FAILURE;
    }
free(input);		/* @1 */
    return rc;
}

static void printUsage(void)
{
    printf("\n");
    printf("sign\n");
    printf("\n");
    printf("Runs TPM_Sign\n");
    printf("\n");
    printf("\t-hk signing key handle\n");
    printf("\t[-pwdk password for signing key (default zeros)]\n");
    printf("\t-if input area to hash and sign\n");
    printf("\t[-os sign signature file name (default do not save)]\n");
    printf("\t[-ik key file name to verify signature (default no verify)]\n");
    printf("\n");
    printf("\t-se0 session handle / attributes\n");
    printf("\t\t01 continue\n");
    exit(1);
}


